### PR TITLE
fix: trade ratio & CoW quote display

### PR DESCRIPTION
**Pull request by 0xApotheosis**
## Description

- Fixes the trade ratio calculation, which was incorrectly handling sell asset fees (which currently only affects CoW Swap quotes)
- Fixes the display of sell asset fees, which were not correctly shown to the user (again, currently only affects CoW Swap quotes)
- Renames `sellAmountBeforeFeesCryptoBaseUnit` to `sellAmountIncludingProtocolFeesCryptoBaseUnit`, because that's what it actually is, and probably caused this bug

This PR is also important as it fixes the ratio data send to MixPanel.

## Pull Request Type

- [x] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [ ] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

N/A

## Risk

Medium - quotes and their ratio/percent better/worse than display might be broken.

## Testing

Get a quote for an asset pair with many supported swappers (e.g. USDC to FOX).
- Ensure that the percent difference looks sane
- Ensure that for CoW Swap quotes, we are subtracting the fee from the receive amount

### Engineering

Inspect CoW quote network requests and ensure that the data matches what we are showing, particularly with regard to fees and receive amount.

### Operations

Check our CoW quotes against https://swap.cow.fi/#/1/swap and ensure they are close.

## Screenshots (if applicable)

Screenshot 2023-08-11 at 4 17 04 pm

### CODE REVIEW

**Comment by 0xApotheosis** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #5062** <a href="https://app.graphite.dev/github/pr/shapeshift/web/5062" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/shapeshift/web/5062?utm_source=stack-comment).

File: src/state/apis/swappers/helpers/getInputOutputRatioFromQuote.ts [hunk 1]
```diff
@@ -94,21 +79,49 @@ const _getReceiveSideAmountsCryptoBaseUnit = ({
...
+          return bnOrZero(firstHopNetworkFee)
+        })()
+      : bn(0)
+  const buySideProtocolFeeCryptoBaseUnit = bnOrZero(buyAssetProtocolFee?.amountCryptoBaseUnit)
 
   const netReceiveAmountCryptoBaseUnit = buyAmountCryptoBaseUnit
     .minus(slippageAmountCryptoBaseUnit)
     .minus(buySideNetworkFeeCryptoBaseUnit)
     .minus(buySideProtocolFeeCryptoBaseUnit)
+    .minus(sellSideProtocolFeeBuyAssetBaseUnit)
```
Comment (0xApotheosis) [inline src/state/apis/swappers/helpers/getInputOutputRatioFromQuote.ts:124]: Pretty much this entire PR is based around this line.

File: src/state/slices/tradeQuoteSlice/helpers.ts [hunk 2]
```diff
@@ -76,14 +77,23 @@ const _getReceiveSideAmountsCryptoBaseUnit = ({
...
+      outputExponent: lastStep.buyAsset.precision,
+    }),
+  ).times(rate)
+  const buySideProtocolFeeCryptoBaseUnit = bnOrZero(buyAssetProtocolFee?.amountCryptoBaseUnit)
 
   const netReceiveAmountCryptoBaseUnit = buyAmountCryptoBaseUnit
     .minus(slippageAmountCryptoBaseUnit)
     .minus(buySideNetworkFeeCryptoBaseUnit)
     .minus(buySideProtocolFeeCryptoBaseUnit)
+    .minus(sellSideProtocolFeeBuyAssetBaseUnit)
```
Comment (0xApotheosis) [inline src/state/slices/tradeQuoteSlice/helpers.ts:96]: Pretty much this entire PR is based around this line.

File: src/components/MultiHopTrade/hooks/useGetTradeQuotes/useGetTradeQuotes.tsx [hunk 3]
```diff
@@ -62,7 +63,10 @@ const getMixPanelDataFromApiQuotes = (quotes: ApiQuote[]): GetMixPanelDataFromAp
     })
     .filter(isSome)
 
-  return { quoteMeta, sellAssetId, buyAssetId, sellAmountUsd }
+  // Add a version string, in the form of an ISO 8601 standard basic format date, to the JSON blob to help with reporting
+  const version = '20230823'
```
Comment (0xApotheosis) [inline src/components/MultiHopTrade/hooks/useGetTradeQuotes/useGetTradeQuotes.tsx:67]: Just needed a unique string I could use to filter the data in Jupyter Notebook.

File: src/state/apis/swappers/helpers/getInputOutputRatioFromQuote.ts [hunk 4]
```diff
@@ -94,21 +79,49 @@ const _getReceiveSideAmountsCryptoBaseUnit = ({
...
+            const otherDenomFee = lastStep.feeData.protocolFees[firstStep.sellAsset.assetId]
+            if (!otherDenomFee) return bn(0)
+            return bnOrZero(otherDenomFee.amountCryptoBaseUnit).times(rate)
+          }
+
+          const firstHopNetworkFee = firstStep.feeData.networkFeeCryptoBaseUnit
+          // Subtract the first-hop network fees for OSMO -> ATOM, which aren't automagically subtracted in the multi-hop abstraction
+          return bnOrZero(firstHopNetworkFee)
+        })()
+      : bn(0)
```
Comment (0xApotheosis) [inline src/state/apis/swappers/helpers/getInputOutputRatioFromQuote.ts:117]: Mostly ignore this change, it's just copied from the duplicate `_getReceiveSideAmountsCryptoBaseUnit` function we have elsewhere, that was not updated when the other one was.

File: src/components/MultiHopTrade/hooks/useGetTradeQuotes/useGetTradeQuotes.tsx (hunk 3 above)
Comment (woodenfurniture) [reply src/components/MultiHopTrade/hooks/useGetTradeQuotes/useGetTradeQuotes.tsx:67]: i thought 8601 was delimited by `-`, e.g `2023-08-23` ?

**Review by woodenfurniture — APPROVED** [review]

File: src/components/MultiHopTrade/hooks/useGetTradeQuotes/useGetTradeQuotes.tsx (hunk 3 above)
Comment (0xApotheosis) [reply src/components/MultiHopTrade/hooks/useGetTradeQuotes/useGetTradeQuotes.tsx:67]: That's the "extended" version. The "normal" version is without the delimiters 😁

**Pull request merged**